test/unit/org/apache/cassandra/io/sstable/format/bti/PartitionIndexTest.java

```diff
@@ -0,0 +1,934 @@
+        try (PartitionIndex index = random.right)
+        {
+            assertEquals(COUNT - 2, index.size());
+            assertEquals(keys.get(1), index.firstKey());
+            assertEquals(keys.get(COUNT - 2), index.lastKey());
+        }
+    }
+
+    public void checkIteration(List<DecoratedKey> keys, int keysSize, PartitionIndex index)
+    {
+        try (PartitionIndex enforceIndexClosing = index;
+             PartitionIndex.IndexPosIterator iter = index.allKeysIterator())
+        {
+            int i = 0;
+            while (true)
+            {
+                long pos = iter.nextIndexPos();
+                if (pos == PartitionIndex.NOT_FOUND)
+                    break;
+                assertEquals(i, pos);
+                ++i;
+            }
+            assertEquals(keysSize, i);
+        }
+    }
+
+    @Test
+    public void testConstrainedIteration() throws IOException
+    {
+        Pair<List<DecoratedKey>, PartitionIndex> random = generateRandomIndex(COUNT);
+        try (PartitionIndex summary = random.right)
+        {
+            List<DecoratedKey> keys = random.left;
+            Random rand = new Random();
+
+            for (int i = 0; i < 1000; ++i)
+            {
+                boolean exactLeft = rand.nextBoolean();
+                boolean exactRight = rand.nextBoolean();
+                DecoratedKey left = exactLeft ? keys.get(rand.nextInt(keys.size())) : generateRandomKey();
+                DecoratedKey right = exactRight ? keys.get(rand.nextInt(keys.size())) : generateRandomKey();
+                if (right.compareTo(left) < 0)
+                {
+                    DecoratedKey t = left;
+                    left = right;
+                    right = t;
+                    boolean b = exactLeft;
+                    exactLeft = exactRight;
+                    exactRight = b;
+                }
+
+                try (PartitionIndex.IndexPosIterator iter = new PartitionIndex.IndexPosIterator(summary, left, right))
+                {
+                    long p = iter.nextIndexPos();
+                    if (p == PartitionIndex.NOT_FOUND)
+                    {
+                        int idx = (int) ge(keys, left); // first greater key
+                        if (idx == -1)
+                            continue;
+                        assertTrue(left + " <= " + keys.get(idx) + " <= " + right + " but " + idx + " wasn't iterated.", right.compareTo(keys.get(idx)) < 0);
+                        continue;
+                    }
+
+                    int idx = (int) p;
+                    if (p > 0)
+                        assertTrue(left.compareTo(keys.get(idx - 1)) > 0);
+                    if (p < keys.size() - 1)
+                        assertTrue(left.compareTo(keys.get(idx + 1)) < 0);
+                    if (exactLeft)      // must be precise on exact, otherwise could be in any relation
+                        assertTrue(left == keys.get(idx));
```

Comment:
```suggestion
                        assertSame(left, keys.get(idx));
```
nit: produces a nicer error I guess

Reply:
Done